cloudslang-compiler/src/main/java/io/cloudslang/lang/compiler/SlangCompilerImpl.java

```diff
@@ -162,6 +194,65 @@ public SystemPropertyModellingResult loadSystemPropertiesFromSource(SlangSource
         return extractProperties(parseModellingResult.getParsedSlang(), source, parseModellingResult.getErrors());
     }
 
+    private void handleStrategyAfterModelling(
+            SlangSource source,
+            PrecompileStrategy precompileStrategy,
+            String filePath,
+            ExecutableModellingResult executableModellingResult) {
+        switch (precompileStrategy) {
+            case WITH_CACHE:
+                cachedPrecompileService.cacheValue(filePath, executableModellingResult, source);
+                break;
+            case WITHOUT_CACHE:
+                break;
+            default:
+                throw new NotImplementedException(generatePreCompileTypeErrorMessage(precompileStrategy));
+        }
+    }
+
+    private CacheResult handleStrategyBeforeModelling(
+            SlangSource source,
+            PrecompileStrategy precompileStrategy,
+            String filePath) {
+        CacheResult cacheResult = null;
+        switch (precompileStrategy) {
+            case WITH_CACHE:
+                cacheResult = cachedPrecompileService.getValueFromCache(filePath, source);
+                break;
+            case WITHOUT_CACHE:
+                break;
+            default:
+                throw new NotImplementedException(generatePreCompileTypeErrorMessage(precompileStrategy));
+        }
+        return cacheResult;
+    }
+
+    private String generatePreCompileTypeErrorMessage(PrecompileStrategy precompileStrategy) {
+        return "Precompile type[" + precompileStrategy + "] not yet implemented";
+    }
+
+    private ExecutableModellingResult preCompileModel(SlangSource source) {
+        //first thing we parse the yaml file into java maps
+        ParsedSlang parsedSlang = yamlParser.parse(source);
+        ParseModellingResult parseModellingResult = yamlParser.validate(parsedSlang);
+
+        // Then we transform the parsed Slang source to a Slang model
+        return slangModeller.createModel(parseModellingResult);
+    }
+
+    private boolean isValidCachedValue(CacheResult cacheResult) {
+        CacheValueState cacheValueState = cacheResult.getState();
+        switch (cacheValueState) {
+            case VALID:
+                return true;
+            case MISSING:
+            case OUTDATED:
+                return false;
+            default:
+                throw new NotImplementedException("Cache value state[" + cacheValueState + "] not yet implemented");
```

Comment:
I would throw `NullPointerExceptionfor ` for `null ``cacheValueState `and NotImplementedExceptionon for not null